### ISSUE TITLE
Centered images produce weird HTML

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#2846]
When I upload an image in the editor and leave 'Align' as 'None', this is the HTML that is correctly output:

``` html
<p><img alt="HTML code" data-file-id="7" src="/files/images/pexels-photo-270360.jpeg" width="1200" height="799"></p>
<p>One of the original iterations of this website had a page explaining how I came to learn web design. As that page no longer exists, I thought I'd re-create it here...</p>
```

However, if I change 'Align' to 'Center', this is what is (I believe, incorrectly) output:

``` html
<p></p>
<div class="centered-wrapper"><img alt="HTML code" data-file-id="7" src="/files/images/pexels-photo-270360.jpeg" class="align-center" width="1200" height="799"></div>
<p>One of the original iterations of this website had a page explaining how I came to learn web design. As that page no longer exists, I thought I'd re-create it here...</p>
```

There's an empty paragraph above the image and, since the image is now outside of the paragraph, it butts up against the following paragraph with no margin between them.

I think adding the "centered-wrapper" class to the paragraph itself should suffice. For example:

``` html
<p class="centered-wrapper"><img alt="HTML code" data-file-id="7" src="/files/images/pexels-photo-270360.jpeg" class="align-center" width="1200" height="799"></p>
<p>One of the original iterations of this website had a page explaining how I came to learn web design. As that page no longer exists, I thought I'd re-create it here...</p>
```

Or if not, then at least remove the empty preceding paragraph and I can style the div with margins accordingly.

**Comment by olafgrabienski** [comment]
That's an interesting issue! I can confirm the behaviour: When you change *Align* to *Center*, you get an empty paragraph followed by a wrapper `div` around the image.

Two observations:

- When I change the editor mode to Source code, the `div` isn't there. It appears only when I view the node.
- The `div` is even rendered if I don't allow this HTML tag in the text format configuration.

**Comment by BWPanda** [comment]
Also, more weirdness if you try to add a link to the centered image (link gets added to empty paragraph AND div).

**Comment by BWPanda** [comment]
I'm submitted a PR that fixes this for me, however it's likely not the best solution and I had trouble with figures (I don't normally use them, and in my testing they seemed to be applied after the centering (so not sure why they're in this bit of code)).

As a comparison, here's the output HTML from my PR:
``` html
<p>Some text and paragraphs here...</p>
<p class="centered-wrapper"><a href="/files/images/crocosmia.jpg"><img alt="" data-file-id="42" src="/files/images/crocosmia.jpg" width="500" /></a></p>
<p>And an image above.</p>
```

It adds the 'centered-wrapper' class to the nearest `<p>` tag instead of adding a new `<div>` and it removes the 'align-center' class from the `<img>` itself (since it's not needed and just causes the image to become a block-level element, which affects the anchor around it).

Modifications and suggestions welcome!

**Comment by olafgrabienski** [comment]
Changed the tags (added "needs review") to get suggestions.

**Comment by quicksketch** [comment]
Thanks @BWPanda for the PR! Your suggestions to the markup sound good.

We encountered the "empty paragraph tag" bug in CKEditor when we added linking to images I believe. It's been fixed in Backdrop in the original commit (https://github.com/quicksketch/backdrop/commit/bf040205d4c009942d5e04044ae0eb9985cc377e) but I made a separate issue in the Drupal queue at https://www.drupal.org/node/2876094.

We probably need a similar fix here for centering images. Apparently if the DOM is modified where the caret is, CKEditor reacts by creating an empty paragraph tag to put the caret in. Refocusing the image after centering may fix the problem.

We'll also want to make the markup in CKEditor match as closely as possible. It's not a perfect match because CKEditor adds its own wrappers for resizing images, but if we're changing the front-end markup, we'll probably have to get into the CKEditor plugin too.

**Comment by olafgrabienski** [comment]
I've encountered a related issue which applies with and without the PR by @BWPanda: Images with a caption produce an empty paragraph above and another one below the image:

```
<h3>Image with caption</h3>
<p></p>
<figure class="caption caption-img">
  <img alt="Dummy image" data-file-id="2" height="450" src="/files/inline-images/dummy-04.png" width="600" />
  <figcaption>This is an image caption.</figcaption>
</figure>
<p></p>
<p>Lorem ipsum dolor sit amet...</p>
```

Adding the "help wanted" tag. (I'm very interested even in a workaround because I've a site running with a paragraph counting script which doesn't produce correct results as is counts the empty paragraphs.)

**Comment by klonos** [comment]
Not sure if this was the case before, but in the recent 1.12 release images cannot be centered in the center. Their captions are centered though.

Should we be supporting different aligning for images and their captions? For example a left-aligned image with a center-aligned caption, or a centre-aligned image with a right-aligned caption.

**Comment by Graham-72** [comment]
@klonos I tried to replicate this. Although the image was not centered in the preview, it was centered after saving.

**Comment by klonos** [comment]
@Graham-72 yep, I was only checking the preview; never actually attempted to save the node.

**Comment by jenlampton** [comment]
> Although the image was not centered in the preview

I'm pretty sure it used to be aligned in the preview. Either way, we should fix that too.

**Comment by jenlampton** [comment]
It looks like the problem with image + caption is not related to this PR. 

In the backdropimagecaption plugin there is a line that adds classes to the `<figure>` tag from `editor.config.image2_captionedClass`. This includes the '.caption' class which is intended for use on the `<caption>` tag only. When this class is added to the `<figure>` tag, the alignment breaks due to CSS for the `<caption>` tag being added to the `<figure>` tag.

I'm going to remove that line and see what happens.

**Comment by quicksketch** [comment]
> In the backdropimagecaption plugin there is a line that adds classes to the `<figure> `tag from editor.config.image2_captionedClass. This includes the '.caption' class which is intended for use on the `<caption>` tag only.

I don't think this is correct. That option _is_ intended to be added to the figure tag. From the docs at https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-image2_captionedClass

> A CSS class applied to the `<figure>` element of a captioned image.

In our case, we use it to distinguish between a `<figure>` element that contains a caption and one that does not contain a caption. So I believe that this functionality should be retained.

**Comment by quicksketch** [comment]
I pushed up a new PR at https://github.com/backdrop/backdrop/pull/2982 that iterates on @BWPanda's PR. There were a few issues that I encountered in testing:

- If the auto-p filter was turned off, no parent `<p>` tag would be found, causing an infinite `while` loop.
- Now if no parent is found at all, a `<p>` tag is generated on the fly.
- If an existing `<p>` tag is found, any classes on that `<p>` tag should be maintained, only with `centered-wrapper` added.

I think this PR addresses all these issues.

**Comment by olafgrabienski** [comment]
> I pushed up a new PR at backdrop/backdrop#2982 ...

Hm, the PR has been closed, accidentally or on purpose?

**Comment by jenlampton** [comment]
@olafgrabienski re-opened

**Comment by jenlampton** [comment]
The code also looks good, but we may have some failing tests.

**Comment by herbdool** [comment]
Code looks good to me. Resolve conflicts with 1.x and it's ready in my opinion.

**Comment by olafgrabienski** [comment]
HTML code looks good in the sandbox for an centered image *without* a caption:

```
<p>Some text, and then a centered image:</p>
<p class="centered-wrapper">
<img alt="My image" data-file-id="1" height="480" src="/files/inline-images/my-image.jpg" width="640" class="align-center">
</p>
<p>More text.</p>
```

Interestingly, for a centered image *with* a caption, I see two empty paragraphs in the HTML:

```
<p>Some text, and then a centered image:</p>
<p></p>
<figure class="caption caption-img align-center">
<img alt="My image" data-file-id="1" height="480" src="/files/inline-images/my-image.jpg" width="640">
<figcaption>This is an image caption</figcaption>
</figure>
<p></p>
<p>More text.</p>
```

**Comment by herbdool** [comment]
Based on @olafgrabienski test needs more work.

**Comment by jenlampton** [comment]
I'm unable to reproduce the issue @olafgrabienski reported, here's my recent test page:
http://2982.backdrop.backdrop.qa.backdropcms.org/jens-test-page

Markup looks like this:
```
<p>first paragraph</p>
<p class="centered-wrapper"><img alt="" data-file-id="3" src="/files/inline-images/Artie-jump.png" width="300" class="align-center" /></p>
<p>third paragraph</p>
```

**Comment by olafgrabienski** [comment]
@jenlampton Thanks for having a look at the issue. The empty paragraphs only appear when you *add a caption* to the image. I've verified this again on a [new test page](http://2982.backdrop.backdrop.qa.backdropcms.org/olafs-test-page). Result:

```
<p>First paragraph</p>
<p></p>
<figure class="caption caption-img align-center"><img alt="Dummy image" data-file-id="4" height="200" src="/files/inline-images/dummy_300.png" width="300">
<figcaption>Figure: Dummy image</figcaption>
</figure>
<p></p>
<p>Third Paragraph</p>
```

Btw, for an image with a caption it doesn't matters if the image is centered or not.

Back to "needs work", I guess.
